enhancements/console/multi-cluster-console.md

```diff
@@ -0,0 +1,366 @@
+API server on any of the spoke clusters in the fleet. API requests from the
+frontend go to the console backend, which in turn proxies each request to the
+public API server URL for the selected cluster. The console backend reads the
+`X-Cluster` request header set by the frontend to determine the cluster to
+route the request to.
+
+![Figure 2: Console backend proxy](console-multicluster-proxy.svg)
+
+The various console hooks and utilities that make requests to the API server
+will be updated to read the current cluster from the Redux store and include
+the right `X-Cluster` header. This lets us implement multi-cluster without
+needing to update every page and component that makes API requests. The
+utilities the will automatically add the `X-Cluster` header include
+
+* The `useK8sWatchResource` and `useK8sWatchResources` hooks
+* The `Firehose` component
+* The `k8s{Get,Create,Update,Patch,Delete}` functions
+
+The console operator watches the `ManagedCluster` custom resources created by
+ACM to determine what clusters are in the fleet. It reads the [API server
+public URL and CA file](https://issues.redhat.com/browse/ACM-786) for each
+`ManagedCluster`, and passes them to the console backend as configuration using
+the existing `console-config` config map in the `openshift-console` namespace.
+On startup, the console backend will set up a proxy to each of the clusters in
+the fleet. When the `ManagedCluster` list changes, the console operator will
+update the console configuration, triggering a new rollout of console.
+
+![Figure 3: Operator watches ManagedClusters](operator-watches-clusters.svg)
+
+### Authentication
+
+The OpenShift console backend must support authentication challenges from each
+of the clusters in the fleet. The backend discovers the OAuth endpoints for
+each `ManagedCluster` through the `/.well-known/oauth-authorization-server`
+endpoint and sets up separate authenticators for each cluster. The user's
+access token for each spoke cluster is be stored in a separate cookie that
+includes the cluster name, `openshift-session-token-<cluster-name>`.
+
+The console operator will read the CA file for the spoke cluster OAuth servers
+using a `ManagedClusterView` and pass the correct CA file to console backend as
+part of the multi-cluster config. For 4.9 clusters, this is available in the
+`oauth-serving-cert` config map in the `openshift-config-managed` namespace:
+
+```yaml
+apiVersion: view.open-cluster-management.io/v1beta1
+kind: ManagedClusterView
+metadata:
+  name: get-oauth-serving-cert
+  namespace: cluster1
+spec:
+  scope:
+    resource: configmaps
+    name: oauth-serving-cert
+    namespace: openshift-config-managed
+```
+
+For 4.8 and earlier, the console operator will need to use the
+`default-ingress-cert` from the `openshift-config-managed` namespace.
+
+```yaml
+apiVersion: view.open-cluster-management.io/v1beta1
+kind: ManagedClusterView
```

Comment:
Just to confirm, this is the CR that console-operator on the spoke cluster will fetch?

Reply:
Console operator creates this on the hub cluster. ACM will set the status of the `ManagedClusterView` resource on the hub based on the data from the spoke.